Add UPnP view and new port scanning API

pablomendezroyo opened the pull request:
### UPnP common issues

UPnP is one of the most critical features in DAppNode. We have poor control over routers, using the library miniUPnPC, since every router is different and its behavior may change from one to another.

This situation generates a lot of support and chaos in users:

- UPnP is disabled, and users do not even how to enable it.
- UPnP is disabled and the router does not even have this property. Users should have something in the UI advertising to them which ports are necessary and for what.
- UPnP is enabled but dappnode cannot open all the necessary ports, so users do not even know which ports need to be manually opened and the status of these ports
- UPnP is enabled and users want to know which ports and their status of those ones managed by UPnP.

**If UPnP is enabled, and any port is manually closed it will automatically be re-opened**

### Main changes

- Used an already created call `getPortStatus` which have useful information about UPnP and their ports status
- Added new view at **support > UPnP**,  there are 3 main possibilities
     - DAppNode running in a remote machine. UPnP is not necessary
     - DAppNode running locally, UPnP detected as enabled
     - DAppNode running locally, UPnP detected as disabled

- Added API call to scan TCP ports and show their status 

**Results**

**I could not recreate the situation of having UpNp enabled and some ports closed, since these ports are automatically opened by dappmanager**

**DAppNode running in remote**
![image](https://user-images.githubusercontent.com/41727368/106449422-3efa8c00-6484-11eb-944c-52414b3e5546.png)


**UpNp enabled and ports succesfully mapped**
![image](https://user-images.githubusercontent.com/41727368/106436055-ede29c00-6473-11eb-9c39-7943402cfe03.png)

**UpNp disabled and ports to be open shown**

![image](https://user-images.githubusercontent.com/41727368/106583908-dffd4b80-6545-11eb-89ce-195631d75054.png)

**Scanning**

![image](https://user-images.githubusercontent.com/41727368/106584228-3d919800-6546-11eb-9593-2b47720268fa.png)

dapplion commented on packages/admin-ui/src/pages/support/components/UPnP.tsx:
```diff
@@ -0,0 +1,165 @@
+  const upnpInfo = useApi.getPortsStatus();
+
+  return (
+    <>
```
It's not necessary to have this initial React.Fragment

dapplion commented on packages/admin-ui/src/pages/support/components/UPnP.tsx:
```diff
@@ -0,0 +1,165 @@
+  const portsScanInfo = useApi.portsScanGet();
+
+  return (
+    <>
```
It's not necessary to have this initial React.Fragment

dapplion commented on packages/dappmanager/src/utils/getPortsScan.ts:
```diff
@@ -0,0 +1,21 @@
+}): Promise<PortsScanResponse> {
+  try {
+    const response = await fetch(
+      `https://mighty-escarpment-70069.herokuapp.com/${publicIp}/${ports.join(
```
Move the url to params.ts as `PORT_SCANNER_SERVICE_URL`

dapplion commented on packages/admin-ui/src/pages/support/components/UPnP.tsx:
```diff
@@ -0,0 +1,165 @@
+
+function PortsToOpen() {
+  const upnpInfo = useApi.getPortsStatus();
+  const portsScanInfo = useApi.portsScanGet();
```
A port scan should be triggered manually by the user, not automatically

dapplion commented on packages/admin-ui/src/pages/support/components/UPnP.tsx:
```diff
@@ -0,0 +1,165 @@
+                      </td>
+                    </tr>
+                  );
+                })}{" "}
```
Unnecessary `" "`

dapplion commented on packages/admin-ui/src/common/types.ts:
```diff
@@ -441,6 +448,14 @@ export type InstalledPackageData = Pick<
+export interface PortsScanResponse {
+  host: string; //"83.32.61.31"
+  ports: {
+    open: number[]; // [8092,4001]
```
Even if we only scan TCP, this information should be clear here. The array should be of type `{port: number, protocol: "TCP" | "UDP" }`

dapplion replied:
It should return an array of all the ports requested with the status as a property, instead of sorted in two arrays.
```ts
ports: {
  port: number;
  protocol: "TCP" | "UDP";
  status: "open" | "closed" | "error";
  error?: string;
}[]
```

dapplion commented on packages/admin-ui/src/pages/support/components/UPnP.tsx:
```diff
@@ -0,0 +1,165 @@
+  );
+}
+
+function PortsToOpen() {
```
I think you could merge `PortsOpened` and `PortsToOpen` into the same component and reduce code duplication

dapplion commented on packages/dappmanager/src/calls/portsScanGet.ts:
```diff
@@ -1,11 +1,11 @@
   const packagePort = db.portsToOpen.get();
   // udp ports cannot be scanned
-  // as a better approach, if tcp ports were opened udp quite probably were opened as well
+  // as a better approach, if tcp ports were opened with UpNp, udp quite probably were opened as well
```
UPnP capitalization

dapplion commented on packages/dappmanager/src/utils/getPortsScan.ts:
```diff
@@ -0,0 +1,22 @@
+
+const apiEndpoint = params.PORT_SCANNER_SERVICE_URL;
+
+export async function getPortsScan({
```
Rename to `performPortsScan`

dapplion commented on packages/dappmanager/src/calls/portsScanGet.ts:
```diff
@@ -0,0 +1,17 @@
+  const packagePort = db.portsToOpen.get();
+  // udp ports cannot be scanned
+  // as a better approach, if tcp ports were opened with UpNp, udp quite probably were opened as well
+  const tcpPorts = packagePort.filter(port => port.protocol === "TCP");
```
This formating should be done in the utility `getPortsScan` to keep things together. This way you can use a fully typed interface instead of `ports: string[]` which actually means only the TCP port number in string form

dapplion commented on packages/dappmanager/src/utils/getPortsScan.ts:
```diff
@@ -0,0 +1,22 @@
+    const response = await fetch(
+      `${apiEndpoint}/${publicIp}?tcpPorts=${ports.join(",")}`
+    );
+    return await response.json();
```
You should sanitize the response from the API before sending it to the UI, just in case

pablomendezroyo commented:
## PR update

- Added API response sanitize.
- Added service name to the port table
- Added advanced mode: show API and UPnP port status
- Added logic to determine when to take API or UPnP port status (use to be API
- Added logic to build the necessary data for the ports table in the backend

![upnpEnabledOn](https://user-images.githubusercontent.com/41727368/107198100-0105e680-69f5-11eb-9aac-68097d256e76.png)
![upnpDisabledOff](https://user-images.githubusercontent.com/41727368/107196343-cef38500-69f2-11eb-8085-e561c1fd0fc4.png)
![upnpDisabledOn](https://user-images.githubusercontent.com/41727368/107196357-d2870c00-69f2-11eb-8593-52a12a08934a.png)
![upnpEnabledOff](https://user-images.githubusercontent.com/41727368/107196370-d581fc80-69f2-11eb-9241-fc90e64b426b.png)

dapplion commented on packages/dappmanager/src/utils/portsTableData.ts:
```diff
@@ -0,0 +1,115 @@
+      protocol: port.protocol,
+      upnpStatus: getUpnpStatus({ port, upnpAvailable, upnpPortMappings }),
+      apiStatus: getApiStatus({ port, apiTcpPortsStatus }),
+      service: getServiceName({ port, packages })
```
You should use strictly one this "name" entities: `dnpName`, `serviceName`, `containerName`. Otherwise things get confusing. Note that `serviceName` does not uniquely identify a service globally. You should return `dnpName`, `serviceName` from the dappmanager and the UI should print that in a pretty way

dapplion commented on packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:
```diff
@@ -1,13 +1,12 @@
 import Switch from "components/Switch";
 import React, { useState } from "react";
+import { useApi } from "api";
 import { Table } from "react-bootstrap";
```
You should never import react-bootstrap components from the root to ensure tree-shaking works well. See https://react-bootstrap.github.io/getting-started/introduction/#importing-components

dapplion commented on packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:
```diff
@@ -42,9 +24,10 @@ export function PortsStatusTable() {
       <SubTitle>
         Ports table
         <Button
+          disabled={upnpInfo.isValidating ? true : false}
```
`isValidating` is already a boolean

pablomendezroyo commented on packages/dappmanager/src/modules/portsTable/getApiStatus.ts:
```diff
@@ -0,0 +1,26 @@
+import { PackagePort, TcpPortScan, ApiStatus } from "../../common";
```
donde, only **PortToOpen** is imported from types in the dappmanager

dapplion commented on packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:
```diff
@@ -0,0 +1,223 @@
+          portsToOpen: portsToOpen.data
+        });
+        setApiScanResult(apiPorts);
+        setApiReqStatus({ result: apiPorts });
```
The result of the api call is stored twice. Remove useState for `setApiScanResult` and `setUpnpScanResult`

dapplion commented on packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:
```diff
@@ -0,0 +1,223 @@
+      }
+  }
+
+  function RenderApiStatus({
```
You should not declare components inside components unless strictly necessary. Move them to the body of the module

dapplion commented on packages/dappmanager/src/modules/portsTable/performPortsScan.ts:
```diff
@@ -0,0 +1,89 @@
+ * Check if the response from the scan ports API
+ * is well formatted
+ */
+function sanitizeApiResponse(responseJson: PortScanResponse): void {
```
@pablomendezroyo Bump

dapplion commented on packages/dappmanager/src/daemons/natRenewal/index.ts:
```diff
@@ -25,6 +26,7 @@ async function natRenewal(): Promise<void> {
     // portMappings = [ {protocol: 'UDP', exPort: '500', inPort: '500'} ]
     try {
       const portMappings = await upnpc.list();
+
```
Unnecessary diff
